compiler/generator_rust.c

```diff
@@ -0,0 +1,1355 @@
+
+    int i = 0;
+    int l = strlen(input);
+
+    while (i < l) {
+        int ch = input[i++];
+        if (ch != '~') {
+            write_char(g, ch);
+            continue;
+        }
+        switch (input[i++]) {
+            default: write_char(g, input[i - 1]); continue;
+            case 'C': write_comment(g, p); continue;
+            case 'f': write_block_start(g);
+                      write_failure(g);
+                      g->unreachable = false;
+                      write_block_end(g);
+                      continue;
+            case 'M': write_margin(g); continue;
+            case 'N': write_newline(g); continue;
+            case '{': write_block_start(g); continue;
+            case '}': write_block_end(g); continue;
+            case 'S': write_string(g, g->S[input[i++] - '0']); continue;
+            case 'B': write_b(g, g->B[input[i++] - '0']); continue;
+            case 'I': write_int(g, g->I[input[i++] - '0']); continue;
+            case 'V': write_varref(g, g->V[input[i++] - '0']); continue;
+            case 'W': write_varname(g, g->V[input[i++] - '0']); continue;
+            case 'L': write_literal_string(g, g->L[input[i++] - '0']); continue;
+            case '+': g->margin++; continue;
+            case '-': g->margin--; continue;
+            case 'n': write_string(g, g->options->name); continue;
+        }
+    }
+}
+
+static void w(struct generator * g, const char * s) {
+    writef(g, s, 0);
+}
+
+static void generate_AE(struct generator * g, struct node * p) {
+    const char * s;
+    switch (p->type) {
+        case c_name:
+            write_varref(g, p->name); break;
+        case c_number:
+            write_int(g, p->number); break;
+        case c_maxint:
+            write_string(g, "sys.maxsize"); break;
+        case c_minint:
+            write_string(g, "(~sys.maxsize)"); break;
+        case c_neg:
+            write_char(g, '-'); generate_AE(g, p->right); break;
+        case c_multiply:
+            s = " * "; goto label0;
+        case c_plus:
+            s = " + "; goto label0;
+        case c_minus:
+            s = " - "; goto label0;
+        case c_divide:
+            s = " / ";
+        label0:
+            write_char(g, '('); generate_AE(g, p->left);
+            write_string(g, s); generate_AE(g, p->right); write_char(g, ')'); break;
+        case c_cursor:
+            w(g, "env.cursor"); break;
+        case c_limit:
+            w(g, p->mode == m_forward ? "env.limit;" : "env.limit_backward;"); break;
+        case c_lenof: 
+            g->V[0] = p->name;
+            w(g, "~V0.chars().count()");
+            break;            
+        case c_sizeof:
+            g->V[0] = p->name;
+            w(g, "~V0.len()");
+            break;
+        case c_len: 
+            w(g, "env.current.chars().count()");
+            break;
+        case c_size:
+            w(g, "env.current.len()");
+            break;
```

Comment:
I think it would indeed requiring storing two versions of every literal string.

Looks like you can just call `get_utf8()` (in `space.c`) in a loop to get the Unicode codepoints from a UTF-8 string.